tensorzero_internal/src/cache.rs

```diff
@@ -0,0 +1,213 @@
+pub struct CacheOptions {
+    #[serde(default)]
+    pub enabled: bool,
+    #[serde(default)]
+    pub lookback_s: Option<u32>,
+}
+
+#[derive(Debug, Clone)]
+pub struct ModelProviderRequest<'request> {
+    pub request: &'request ModelInferenceRequest<'request>,
+    pub model_name: &'request str,
+    pub provider_name: &'request str,
+}
+
+impl ModelProviderRequest<'_> {
+    pub fn get_cache_key(&self) -> Result<[u8; 32], Error> {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(self.model_name.as_bytes());
+        hasher.update(self.provider_name.as_bytes());
+        let serialized_request = serde_json::to_string(self.request).map_err(|e| {
+            Error::new(ErrorDetails::Serialization {
+                message: format!("Failed to serialize request: {e}"),
+            })
+        })?;
+        let request_bytes = serialized_request.as_bytes();
+        hasher.update(request_bytes);
+        Ok(*hasher.finalize().as_bytes())
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct ModelInferenceCacheRow {
+    short_cache_key: u64,
+    long_cache_key: String,
+    output: Vec<ContentBlock>,
+    raw_request: String,
+    raw_response: String,
+}
+
+// This doesn't block
+pub fn start_cache_write(
+    clickhouse_client: &ClickHouseConnectionInfo,
+    request: ModelProviderRequest<'_>,
+    output: &[ContentBlock],
+    raw_request: &str,
+    raw_response: &str,
+) -> Result<(), Error> {
+    let cache_key = request.get_cache_key()?;
+    let short_cache_key = u64::from_le_bytes(cache_key[..8].try_into().map_err(|e| {
+        Error::new(ErrorDetails::Cache {
+            message: format!("failed to convert hash into u64 for short cache key: {e}"),
+        })
+    })?);
+    let long_cache_key = hex::encode(cache_key);
+    let output = output.to_owned();
+    let raw_request = raw_request.to_string();
+    let raw_response = raw_response.to_string();
+    let clickhouse_client = clickhouse_client.clone();
+    tokio::spawn(async move {
+        clickhouse_client
+            .write(
+                &[ModelInferenceCacheRow {
+                    short_cache_key,
+                    long_cache_key,
+                    output,
+                    raw_request,
+                    raw_response,
+                }],
+                "ModelInferenceCache",
+            )
+            .await
+    });
+    Ok(())
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CacheLookupResult {
+    #[serde(deserialize_with = "deserialize_json_string")]
+    pub output: Vec<ContentBlock>,
+    pub raw_request: String,
+    pub raw_response: String,
+}
+
+pub async fn cache_lookup(
+    clickhouse_connection_info: &ClickHouseConnectionInfo,
+    request: ModelProviderRequest<'_>,
+    lookback_s: Option<u32>,
+) -> Result<Option<ModelInferenceResponse>, Error> {
+    let cache_key = request.get_cache_key()?;
+    let short_cache_key = u64::from_le_bytes(cache_key[..8].try_into().map_err(|e| {
```

Comment:
Could you add a comment noting that this is just a performance optimization, and that we always check against the long cache key before returning a response?